rllib/core/torch/torch_sarl_trainer.py

```diff
@@ -0,0 +1,172 @@
+import numpy as np
+import torch
+from torch.nn.parallel import DistributedDataParallel as DDP
+
+from ray.rllib.core.sarl_trainer import SARLTrainer
+
+from ray.util.queue import Queue
+
+import ray.train as train
+from ray.air import session
+from ray.air.config import ScalingConfig
+from ray.train._internal.utils import construct_train_func
+from ray.train._internal.backend_executor import BackendExecutor
+from ray.train.torch import TorchConfig
+from ray.train import TrainingIterator
+from ray.train._internal.checkpoint import CheckpointManager
+from ray.train._internal.dataset_spec import RayDatasetSpec
+
+
+class TorchSARLTrainer(SARLTrainer):
+    def __init__(self, config):
+        num_workers = config.get("num_gpus", 0) or 1
+        use_gpu = bool(config.get("num_gpus", 0))
+        init_rl_module_fn = config.get("rl_module_init_fn", None)
+        module_config = config.get("rl_module_config", {})
+        """
+        we can't use fractional gpus with this :(
+
+        Leads to nccl errors:
+
+        ncclInvalidUsage: This usually reflects invalid usage of NCCL library
+        (such as too many async ops, too many collectives at once, mixing
+        streams in a group, etc).
+        """
+        scaling_config = ScalingConfig(
+            num_workers=num_workers,
+            use_gpu=use_gpu,
+        )
+        self.queues, self.training_iterator = self._make_ray_train_trainer(
+            scaling_config, init_rl_module_fn, module_config
+        )
+
+    @staticmethod
+    def init_rl_module(module_config):
+        raise NotImplementedError
+
+    @staticmethod
+    def init_optimizer(module, optimizer_config):
+        raise NotImplementedError
+
+    @staticmethod
+    def compute_loss_and_update(module, batch, optimizer):
```

Comment:
What is the difference between `update` and `compute_grads_and_apply`? Intuitively, I would think `update` updates the model weights, but then what does `compute_...` do?

Reply:
`train` / `update` are user facing methods that the user calls on a `TorchSARLTrainer` instance. `compute_grads_and_apply` is called by ray train workers